### PR TITLE
♻️ make player api declarative

### PR DESCRIPTION
**Pull request by anuejn**
this moves complexity into the player by giving it just the wanted state. this simplifies redux reducer code a lot.

### CODE REVIEW

**Comment by anuejn** [comment]
this should close #142 by not making it nescessary to supress such things anymore. they are now correctly handeled.

File: app/src/core/player.ts [hunk 1]
```diff
@@ -1,71 +1,151 @@
...
+  getCurrentRenderItem(): RenderItem | undefined {
+    return this.renderItems.find(
+      (item) =>
+        item.absoluteStart <= this.currentTime &&
+        item.absoluteStart + item.length >= this.currentTime
+    );
+  }
+
+  play(): void {
+    const lastRenderItem = this.renderItems[this.renderItems.length - 1];
```
Comment (pajowu) [inline app/src/core/player.ts:90]: maybe move to function

File: app/src/core/player.ts [hunk 2]
```diff
@@ -1,71 +1,151 @@
...
+          this.play();
+        } else {
+          requestAnimationFrame(callback);
+        }
+      };
+      callback();
+      element.play();
+    } else {
+      // we produce synthetic silence
+      const startTime = +new Date();
```
Comment (pajowu) [inline app/src/core/player.ts:129]: please comment on `+new Date()`

File: app/src/core/player.ts [hunk 3]
```diff
@@ -1,71 +1,151 @@
...
     }
+
+    const currentRenderItem = this.getCurrentRenderItem();
+    assertSome(currentRenderItem);
+
+    const offset = this.currentTime - currentRenderItem.absoluteStart;
+    if ('source' in currentRenderItem && currentRenderItem.source) {
+      const element = this.sources[currentRenderItem.source];
+      element.currentTime = currentRenderItem.sourceStart + offset;
+      const callback = () => {
```
Comment (pajowu) [inline app/src/core/player.ts:104]: can we deduplicate the callbacks for this and artificial silence

**Review by pajowu — CHANGES_REQUESTED** [review]

**Review by pajowu — APPROVED** [review]

**Pull request merged**